src/EditorFeatures/Core/Diagnostics/RawDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,165 @@
+
+    /// <summary>
+    /// When we hear about a new event cancel the costly work we're doing and compute against the latest snapshot.
+    /// </summary>
+    protected sealed override bool CancelOnNewWork => true;
+
+    protected sealed override bool TagEquals(TTag tag1, TTag tag2)
+        => _callback.TagEquals(tag1, tag2);
+
+    protected sealed override ITaggerEventSource CreateEventSource(ITextView? textView, ITextBuffer subjectBuffer)
+    {
+        // OnTextChanged is added for diagnostics in source generated files: it's possible that the analyzer driver
+        // executed on content which was produced by a source generator but is not yet reflected in an open text
+        // buffer for that generated file. In this case, we need to update the tags after the buffer updates (which
+        // triggers a text changed event) to ensure diagnostics are positioned correctly.
+        return TaggerEventSources.Compose(
+            TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer),
+            TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer),
+            TaggerEventSources.OnDiagnosticsChanged(subjectBuffer, _diagnosticService),
+            TaggerEventSources.OnTextChanged(subjectBuffer));
+    }
+
+    protected sealed override Task ProduceTagsAsync(
+        TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, int? caretPosition, CancellationToken cancellationToken)
+    {
+        return ProduceTagsAsync(context, spanToTag, cancellationToken);
+    }
+
+    private async Task ProduceTagsAsync(
+        TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, CancellationToken cancellationToken)
+    {
+        if (!_callback.IsEnabled)
+            return;
+
+        var diagnosticMode = GlobalOptions.GetDiagnosticMode(InternalDiagnosticsOptions.NormalDiagnosticMode);
+        if (!_callback.SupportsDiagnosticMode(diagnosticMode))
+            return;
+
+        var document = spanToTag.Document;
+        if (document == null)
+            return;
+
+        var snapshot = spanToTag.SnapshotSpan.Snapshot;
+
+        var workspace = document.Project.Solution.Workspace;
+
+        // See if we've marked any spans as those we want to suppress diagnostics for.
+        // This can happen for buffers used in the preview workspace where some feature
+        // is generating code that it doesn't want errors shown for.
+        var buffer = snapshot.TextBuffer;
+        var suppressedDiagnosticsSpans = (NormalizedSnapshotSpanCollection?)null;
+        buffer?.Properties.TryGetProperty(PredefinedPreviewTaggerKeys.SuppressDiagnosticsSpansKey, out suppressedDiagnosticsSpans);
+
+        var sourceText = snapshot.AsText();
+
+        try
+        {
+            var diagnostics = await _analyzerService.GetDiagnosticsForSpanAsync(
+                document, range: null, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
right now we're getting all diagnostics. we need an api to get the right syntax/semantic analyzer/compiler.  @mavasani  @dibarbet .

Also, we could pass in the range here.  but it's only worthwhile if it will genuinely mean less computation happens.  e.g. if it still does all computation but just filters the results, then there's no point.

Reply:
> but it's only worthwhile if it will genuinely mean less computation happens

This API already guarantees that we will execute analyzer callbacks only for the nodes/operations/symbols touching the span for analyzers that we know for sure can perform semantic span based analysis (i.e. only our internal built-in analyzers with [DiagnosticCategory.SemanticSpanAnalysis](https://sourceroslyn.io/Microsoft.CodeAnalysis.Workspaces/R/9e3d7dbd159cb5ab.html): https://github.com/dotnet/roslyn/blob/cf9118002332fc5261198e1708be8197ae21f503/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs#L210-L213

Reply:
One of the LSP diagnostics performance improvements that I have discussed with @dibarbet in past is actually having the editor pass in the visible document span for the visible part of the active document. We can than avoid running hidden and unnecessary diagnostic analyzers for non-visible parts of the document for computing tags and also active document analysis. This can help improve the performance for some of our most expensive built in analyzers (simplify type name, remove unnecessary cast, remove unnecessary suppressions)